Bug fixes to allow completion of Hai Reveal campaign

UnorderedSigh opened the pull request:
With these changes, the Hai Reveal campaign should be playable from start to finish. Some of these bugs made it unplayable, others caused glitches, and there are a few textual corrections. Thanks to @MasterOfGrey who provided some fixes and a detailed review, and to @NomadicVolcano and @Hurleveur for advice.

There are four problems I didn't fix, which cause parts of the campaign to not work as intended, but it does still work.

## Testing Done

Played entire game from the beginning, passing through Free Worlds Reconciliation, Hai Reveal, to the beginning of the Wanderers. Also, I obtained Coalition and Remnant outfits licenses to test new ship designs. I did this without meeting the Hai until the message.

For my third set of fixes, I replayed a few more ways:

1. Meet the Hai but  not the Unfettered, and start the campaign.
2. Meet both Hai factions, and start the campaign.
3. Go through the red wormhole. This bypasses a long sequence of missions.

Also, I ran all of the hai reveal files through a spell checker.

## Fix Details

### Things Not Fixed

These problems cause the campaign to not work as intended, but you can still play through the campaign:

1. Hai Reveal [A14] A Time of Trouble - escort randomly-chosen ships that may not have enough fuel. This seams reasonable to me; this is *not* an early game escort mission. Helping a ship with too little fuel is an okay added challenge. 
2. Pincer Beasts cannot keep up with asteroids in some systems. They just wiggle around trying to keep up until they get bored and leave. It's very cute, but (according to NomadicVolcano) it was not the intent. Reverse thrusters and stronger forward thrusters would help.
3. Hai Reveal [B02] Join the Fleet - fails when a ship with "npc accompany" dies, even if it does not have "save." Something in the engine isn't working how it should. My workaround is to remove "accompany" until someone fixes this.
4. Hai Reveal [A07] Deliver the Relays - relays remain in your cargo hold, so you fail if they don't return to Earth. I contrived an excuse: you're returning the deployment hardware.

The "Remnant: Defense 3" mission also deploys satellites, so I'm pinging @Zitchas who may have a solution.

### First Batch of Fixes

The initial fixes before the PR was reviewed:

1. Hai Reveal: Reputation Reset - change Philosophers government to Elenctic Commune
2. Yeertle Hint - wrong type of `"` led to mission description being incomplete
3. hai_reveal_2_intro - three missions used ": Offered" instead of ": offered" for variable names, causing missions to never complete
5. Hai Building Mountaintop [2] Recruits - removed angle braces
6. Hai Reveal [A03a] Symbol Identification - removed angle braces
7. Hai Reveal [A04] Report to Teeneep - mission was never offered due to incorrect source and missing `event: ` in the "to offer" block
8. Hai Reveal [A05] Take Teeneep to Hai-home - mission was never offered due to missing `event: ` in the "to offer" block
9. Hai Reveal [A06a] The Calm - grammar error
10. Hai Reveal [A07] Deliver the Relays description - changed an incorrect/confusing mission description
11. Hai Reveal [A07] Deliver the Relays conversation - removed angle brackets and corrected indentation
12. Hai Reveal [A08] Symbol Identification - proper nouns should be capitalized. Also, this is consistent with other places the Navy Intelligence Observatory is mentioned
13. Hai Reveal [A08] Symbol Identification - paragraph missing initial double quote
14. Hai Reveal [A10] Demanding Satisfaction - extra space before period at end of sentence
15. Hai Reveal [A10] Demanding Satisfaction - corrected typo: "will" should be "well"
16. Hai Reveal [A12] Into the Lion's Den - bug in description: `<stopover>` instead of `<stopovers>`
17. Hai Reveal [A13-A] Unfettered Pickup - ships started in the wrong system
18. Hai Reveal [B01] Information Blackout - making the mission invisible prevented the player from knowing what to do next. Also, mission had the wrong completion condition.
19. Hai Reveal [B01-A] The Blockade Goes Up - corrected the confusing description
20. Hai Reveal [B01-A] The Blockade Goes Up - increased to 12 turns to allow more time for player to realize the ships are starting in distant systems and will need to refuel
21. Hai Reveal [B01-A] The Blockade Goes Up - incorrect offer condition
22. Hai Reveal [B01-A] The Blockade Goes Up - syndicate escorts were missing the "accompany" tag
23. Hai Reveal [B01-A] The Blockade Goes Up - gave more instructions when entering systems
24. Hai Reveal [B01-A] The Blockade Goes Up - added an "on visit" condition to inform player not all defending ships were destroyed
25. Hai Reveal [B03] Contact Giti - corrected typo: "too" instead of "to"
26. Hai Reveal [B04] The Road Not Taken - corrected paragraph indentation (must use tabs, not spaces)
27. Hai Reveal [B05-long] The Red Mystery - better description
28. Hai Reveal [B05-long] The Red Mystery - no completion condition, to ensure description remains until no longer relevant
29. Hai Reveal [B06-long] Tarazed Analysis - added a hint on where ships arrive most frequently
30. event "devil-run smugglers visible" - smugglers arrive more often. (Originally, the campaign had lots of ships in between far north and syndicate near the wormhole. Without them, there are few smugglers to find due to infrequent arrival times. This is confusing to players that don't know what to look for.)
31. Hai Reveal [B07-long] Learn of Keystones - corrected source and added a missing "fail" clause
32. Hai Reveal [B09] Scanning Devil-Hide - grammar error
33. Hai Reveal [B11] Moving Together - grammar error
34. Hai Reveal [B11] Moving Together - added an "on visit" to avoid confusing player
35. Hai Reveal [B12] Devil-Hide Battle - some ships lacked quantum keystones, so they could not reach Devil's Hide
36. Hai Reveal [B12] Devil-Hide Battle - More durable "npc save" ship designs. Ships that had to survive the battle could not reliably do so, due to too little shields. In particular, the Auxiliary, Dreadnought and Centipede frequently died. I tweaked their outfits and played through the battle a few times. If the player has a strong ship that does not use explosive ammo, the battle is quite playable.
37. Hai Reveal [B12] Devil-Hide Battle - autosave, and failure message. This is a hard battle if you don't know what to expect.
38. Hai Reveal [B13-G] Elenctic Guard - changed an incorrect mission description

### More Fixes

After resolving review comments and retesting, I found and fixed a few more bugs:

1. Hai Rescue: Yeertle Hint - does not vanish when Hai Leaks Response 6 cease to be active
2. Hai Rescue: Oonem Hint - does not vanish when Hai Leaks Response 6 cease to be active
3. Hai Rescue: Prison RP Hint - does not vanish when Hai Leaks Response 6 cease to be active
5. Hai Rescue: Meroot and Lucy Hint - does not vanish when Hai Leaks Response 6 cease to be active
6. Hai Reveal [B01-A] The Blockade Goes Up - does not always see the completion of A14, so I had to change its "to offer." The condition is functionally equivalent.
7. Hai Reveal [B02] Join the Fleet - "npc accompany" implies "npc save" so the mission is impossible if any of the escorts are killed. It is expected that some will die in the battle.
8. Hai Reveal [B02] Join the Fleet - "escort waiting" does nothing; it is the same as "escort." There is no way to make the NPC fleets wait to rendezvous with you at a specified system.
9. Hai Reveal [B02] Join the Fleet - fleets start far away, so they run out of fuel before they arrive. I've changed the mission so the fleets all start at Sunracer, in orbit of the player's planet.

### Yet More Fixes

A third set, after another round of review:

1. Better fixes from @MasterOfGrey for hai kidnapping victims missions.
2. Hai Reveal [A07] Deliver the Relays - workaround for a bug I can't fix. The relays remain in the ship after they're "deployed" so the mission does not complete if the escort with the (now deployed) relays hasn't arrived. Now the mission description tells you to return the "deployment hardware" and I've left a note in a comment that we need a better fix.
3. Hai Reveal [B01] Information Blackout - reminder to visit the spaceport remains, after Master of Grey approved me putting it back in.
5. Hai Reveal [B02] Join the Fleet - corrected mission name
6. Hai Reveal [B02] Join the Fleet - ships spawn at Rajak so they have enough fuel to reach Stormhold. There is a comment explaining this.
7. Hai Reveal [B02] Join the Fleet - player is told to meet fleet at Moktar. This ensures the fleet doesn't pass the player and have to jump back, running out of fuel. 
8. Hai Reveal [B02] Join the Fleet - the "wow that's a big fleet" message is tweaked to make sense if the fleet hasn't arrived yet, since the prior sentence says you see their transpoder signals
9. "Thankyou" is not a word
10. Hai Reveal [B06-long A] Smuggler Ambush - mission is able to trigger now
11. Hai Reveal [B06-long] Tarazed Analysis - other half of bug fix to B06-long A 
11. Hai Building Mountaintop [2] Recruits - "human habited" should be "human-inhabited." The word "habited" means "wearing a habit," which is a type of gown.
12. Hai Reveal [B02] Join the Fleet - "smugglings" isn't a word
13. Hai Reveal [C04] Cleanup comment - missing one "d" from great-granddaughter,

## Save Files

There are many steps in this campaign, so I've provided saves at only a few key points. You start with some money, a supply of outfits at New Tortuga, and a few ships parked in various places. Most escorts use new Korath and Coalition outfits.

For the purists: I did not cheat in any ships or missions into save files. All ships were obtained through capture, except the two gift ships from the remnant. Outfits were bought or pillaged, except gifted Cloaking Device and one of the Jump Drives. These bug fixes would have arrived a week earlier if I had used such shortcuts, and not gotten distracted with designing new ships.

- [Friction Diction~3024-03-17 human campaign almost done.txt](https://github.com/endless-sky/endless-sky/files/9848947/Friction.Diction.3024-03-17.human.campaign.almost.done.txt)
- [Friction Diction~3026-03-31 before election results (fewer ships).txt](https://github.com/endless-sky/endless-sky/files/9848976/Friction.Diction.3026-03-31.before.election.results.fewer.ships.txt)
- [Friction Diction~3026-04-11 before part 5 (few ships).txt](https://github.com/endless-sky/endless-sky/files/9848978/Friction.Diction.3026-04-11.before.part.5.few.ships.txt)
- [Friction Diction~3026-04-29 before visiting devil run.txt](https://github.com/endless-sky/endless-sky/files/9848979/Friction.Diction.3026-04-29.before.visiting.devil.run.txt)
- [Friction Diction~3026-05-08 before smugglers.txt](https://github.com/endless-sky/endless-sky/files/9848981/Friction.Diction.3026-05-08.before.smugglers.txt)
- [Friction Diction~3026-07-05 before reporting.txt](https://github.com/endless-sky/endless-sky/files/9848984/Friction.Diction.3026-07-05.before.reporting.txt)
- [Friction Diction~3026-07-26 after devil hide.txt](https://github.com/endless-sky/endless-sky/files/9848986/Friction.Diction.3026-07-26.after.devil.hide.txt)
- [Friction Diction~3026-12-10 after meeting wanderers.txt](https://github.com/endless-sky/endless-sky/files/9848989/Friction.Diction.3026-12-10.after.meeting.wanderers.txt)

MasterOfGrey commented:
Holy mother of jesus. Alrighty, will have a look through this. We knew there'd be things after doing a huge rework to it on a short timeline but wow.

A couple of these are already noted with fixes PR'd, so you might find that some conflicts arise in future - in which case I'd favour the fixes already PR'd. 
This should cover off a lot of stuff though.

MasterOfGrey commented on data/hai/hai reveal 2 start.txt:
```diff
@@ -51,7 +51,7 @@
 	landing
 	name "Investigate a Were-Squirrel"
-	description "A truly outlandish rumor has been floating around that there might be a "were-squirrel" (Hai) hiding out on <destination>."
+	description "A truly outlandish rumor has been floating around that there might be a “were-squirrel” (Hai) hiding out on <destination>."
```
I believe this is how it actually needs to be done because fancy quotes aren't supported.
```suggestion
	description `A truly outlandish rumor has been floating around that there might be a "were-squirrel" (Hai) hiding out on <destination>.`
```

MasterOfGrey commented on data/hai/hai reveal 4 middle.txt:
```diff
@@ -181,7 +181,7 @@ mission "Hai Reveal [A03a] Symbol Identification"
 		conversation
 			`You make what is beginning to feel like a standard request to land at the secure spaceport and disembark with relevant copies of files on a data stick in hand. It doesn't take long to track down Alondo, though you do end up having to wait for nearly half an hour while he finishes up some kind of meeting with a Hai official of some kind. When he's finished he waves you into the room gestures f ...
-			`	<You have a conversation where you show him the symbol, and ask him for leads. He says he knows some people who can be discreet and he'll see if he can find someone who might know something. Also something about him working with Remington lately.>`
+			`	You have a conversation where you show him the symbol, and ask him for leads. He says he knows some people who can be discreet and he'll see if he can find someone who might know something. Also something about him working with Remington lately.`
```
This is actually kind of intentional, this line is on the to-do list to actually have that conversation written up.

UnorderedSigh replied:
I want this to be a bugfix-only PR, so I won't add actual content here. Instead, I'll just make the filler less glaring and add a "TODO" comment above it.

MasterOfGrey commented on data/hai/hai reveal 4 middle.txt:
```diff
@@ -688,10 +688,13 @@ mission "Hai Reveal [A04] Report to Teeneep"
 	name "Report to Teeneep"
 	description `Tell Teeneep about the experience with Scar's Legion.`
-	source Hai-home
+	source
+		government "Hai"
+		not planet Darkwaste
+		not attributes "uninhabited"
```
This isn't necessary. While it feels like logically it should be able to fire anywhere, the previous thing ends on Hai-home so this should just trigger immediately after and save on unnecessary lines.

UnorderedSigh replied:
I think you're right, but I'll retest without this to make sure.

UnorderedSigh replied:
This needs "not planet Darkwaste" since you have to fly to Darkwaste. It does not need the other conditions.

MasterOfGrey commented on data/hai/hai reveal 4 middle.txt:
```diff
@@ -921,7 +924,7 @@ mission "Hai Reveal [A07] Deliver the Relays"
 			`	The man grins and shakes his head as she departs. "Remarkable woman. Anyway, the name's David Joseph Turner, venture capitalist, entrepreneur, and famous Paradise Worlds businessman." Turner grabs a business card from his pocket and hands it to you. The card contains his contact information, an address on Martini, and makes sure to boast in golden letters his net worth of more than three bil ...
 			`	"You may not have heard of me, but I've sure as hell heard of you." Turner turns around and waves his arm across the crowd of people working away on the station concourse. "To be involved in funding and producing this movie though, what an adventure. You know I've paid for movies to be made, or at least partially invested in them, but I've never had to actually be there while it was happenin ...
-			`<Turner takes a moment to enthuse about the movie and relay a brief synopsis to demonstrate how well the production is in fact going.>`
+			`	Turner takes a moment to enthuse about the movie and relay a brief synopsis to demonstrate how well the production is in fact going.`
```
Also a to-do list item to actually insert the conversation, but I suppose this one is fair enough.

UnorderedSigh replied:
I'll add a "TODO" comment above it to make that clear.

MasterOfGrey commented on data/hai/hai reveal 4 middle.txt:
```diff
@@ -1269,7 +1272,7 @@ mission "Hai Reveal [A12] Into the Lion's Den"
 	landing
 	name "A New Elder"
-	description "Travel with Teeneep and the diplomatic corps to <stopover> and escort the Unfettered to <planet>."
+	description "Travel with Teeneep and the diplomatic corps to <stopovers> and escort the Unfettered to <planet>."
```
Did this break in-game?
The wiki does say it's explicitly `<stopovers>` so if it did this should be a wiki update.
![image](https://user-images.githubusercontent.com/60949828/197459115-e5789dd9-5839-40e5-b7cb-a3d42496be23.png)

Hurleveur replied:
<stopover> just shows as <stopover> in game indeed, it needs to be <stopover**s**>

MasterOfGrey commented on data/hai/hai reveal 5 philosophers.txt:
```diff
@@ -13,13 +13,14 @@
 
 mission "Hai Reveal [B01] Information Blackout"
 	landing
-	invisible
 	name "Information Blackout"
 	description "Meet the Syndicate Security Captain at the spaceport on <origin> when you're ready."
 	source "Sunracer"
 	clearance
 	to offer
 		has "Hai Reveal [A14] A Time of Trouble: done"
+	to complete
+		has "Hai Reveal [B02] Join the Fleet: offered"
```
This whole thing is a no - what actually should happen here is that the conversation should end with a `decline` not an accept.

UnorderedSigh replied:
You need to keep a reminder to arrive at the spaceport if the player decides to delay the mission.

Hurleveur replied:
Why would you have a name and description if the mission is not to be shown, that's weird indeed

UnorderedSigh replied:
I've removed those changes.

MasterOfGrey commented on data/hai/hai reveal 5 philosophers.txt:
```diff
@@ -84,16 +85,15 @@ mission "Hai Reveal [B01-A] The Blockade Goes Up"
-	description "Join the fleet in <waypoint> and move on Stormhold."
-	deadline 9
-	# The minimum journey time is 7
+	description "Join the fleet and move on Stormhold. Make sure all ships have enough fuel to reach there at the same time."
+	deadline 12
+	# The minimum journey time is 7 if the player jumps system by system.
 	source "Sunracer"
-	waypoint "Sumar"
 	destination "Stormhold"
-	clearance
 	to offer
-		has "Hai Reveal [B01] Information Blackout: offered"
+		has "Hai Reveal [A14] A Time of Trouble: done"
```
This too is a large amount of no - largely based on the actual changes needed for the previous one.

UnorderedSigh replied:
The mission won't work without clearance, so I should not have taken that out. I was able to finish anyway since I had already dominated the planet.

UnorderedSigh replied:
I changed how I fixed this mission.

MasterOfGrey commented on data/hai/hai reveal 5 philosophers.txt:
```diff
@@ -151,16 +151,22 @@ mission "Hai Reveal [B02] Join the Fleet"
 		fail "Hai Reveal [A11-B] Permanent Fleets"
 	on enter "Sumar"
 		dialog `You have reached the rendezvous and various fleets are arriving. You have never seen so many military vessels in one place. As promised, both Hai and human ships have tied their navigation to your systems for coordinating the next couple of jumps to Alcyone.`
+	on enter "Moktar"
+		dialog `Some of your escorts are traveling from far away, and could run out of fuel before Stormhold. Make sure you refuel them before traveling onward.`
```
If there's a problem with the escorts running out of fuel, then the variants used need more fuel. You are not meant to stop or land anywhere in this mission between departing Sunracer and landing on Stormhold.

UnorderedSigh replied:
Then the escorts should start at your planet, rather than all over the galaxy.

UnorderedSigh replied:
If the escorts start where you are, it fixes most of the problems in the mission sequence. I will make that change and retest.

UnorderedSigh replied:
I see what you were trying to do with "escort waiting." The escorts don't wait. They'll fly to you, just like any other escorts, and run out of fuel. 

I'm reworking the mission to have the escorts start where you are (Sunracer), just like they do in all other missions.  Then the fuel supply won't matter.

MasterOfGrey commented on data/hai/hai reveal 5 philosophers.txt:
```diff
@@ -376,14 +382,14 @@ mission "Hai Reveal [B05-long] The Red Mystery"
 		has "Hai Reveal [B05-long] The Red Mystery: done"
 	to complete
-		has "Hai Reveal [B06-long] Boarded Smuggler: offered"
+		never
```
The way this mission works has just been fixed elsewhere.

UnorderedSigh replied:
You fixed this after I tested the campaign, so I'll use your fixes and retest.

MasterOfGrey commented on data/hai/hai reveal 5 philosophers.txt:
```diff
@@ -376,14 +382,14 @@ mission "Hai Reveal [B05-long] The Red Mystery"
 mission "Hai Reveal [B06-long] Tarazed Analysis"
 	landing
 	name "Seeking Aid"
-	description `Find and board a smugglers ship. Your display will identify such ships as belonging to the "Smuggler" government. Try Syndicate space or the Far North. Head to <planet> afterwards.`
+	description `Board smuggler ships to gain information. Your display will identify such ships as belonging to the "Smuggler" government. Try near Devil's Run in Syndicate space, or the Far North.`
```
The first half, maybe, the second half, no.

UnorderedSigh replied:
The mystery of the outcome of boarding a ship is more interesting, so I'll revert this change.

MasterOfGrey commented on data/hai/hai reveal 5 philosophers.txt:
```diff
@@ -421,6 +427,7 @@ mission "Hai Reveal [B06-long] Tarazed Analysis"
 			choice
 				`	(Return to your ship.)`
 			`	After one last ride through the metro system of <origin>, you arrive at your ship. You plug the data card into your computer, and a new "government" shows up: Smuggler. The ships will be colored differently than pirates.`
+			`	After a moment's thought, you realize that if the smuggling involves Devil's Run, you may find the most smugglers in systems near there.`
```
Redundant with earlier direction, and also no.
This is the B-plotline if the player hasn't actually discovered how to access a wormhole like this before - which is normally something that requires engagement/thought and effort.
It is not meant to spoonfeed the player here.

UnorderedSigh replied:
Agreed. Will remove.

MasterOfGrey commented on data/hai/hai reveal 5 philosophers.txt:
```diff
@@ -448,39 +455,39 @@ event "devil-run smugglers visible"
-		add fleet "Smuggler (Hai Traffickers)" 10000
+		add fleet "Smuggler (Hai Traffickers)" 8000
 	system "Mintaka"
-		add fleet "Smuggler (Hai Traffickers)" 5000
+		add fleet "Smuggler (Hai Traffickers)" 4000
 	system "Gorvi"
-		add fleet "Smuggler (Hai Traffickers)" 5000
+		add fleet "Smuggler (Hai Traffickers)" 4000
 	system "Tortor"
-		add fleet "Smuggler (Hai Traffickers)" 5000
+		add fleet "Smuggler (Hai Traffickers)" 4000
 	system "Unagi"
-		add fleet "Smuggler (Hai Traffickers)" 7000
+		add fleet "Smuggler (Hai Traffickers)" 6000
 	system "Hatysa"
-		add fleet "Smuggler (Hai Traffickers)" 5000
+		add fleet "Smuggler (Hai Traffickers)" 4000
 	
 	# Route to Almach
 	system "Zaurak"
-		add fleet "Smuggler (Hai Traffickers)" 7000
+		add fleet "Smuggler (Hai Traffickers)" 5500
 	system "Sheratan"
-		add fleet "Smuggler (Hai Traffickers)" 7000
+		add fleet "Smuggler (Hai Traffickers)" 5500
 	system "Ruchbah"
-		add fleet "Smuggler (Hai Traffickers)" 7000
+		add fleet "Smuggler (Hai Traffickers)" 5500
 	system "Alpheratz"
-		add fleet "Smuggler (Hai Traffickers)" 7000
+		add fleet "Smuggler (Hai Traffickers)" 5500
 	system "Achernar"
-		add fleet "Smuggler (Hai Traffickers)" 7000
+		add fleet "Smuggler (Hai Traffickers)" 5500
 	system "Al Dhanab"
-		add fleet "Smuggler (Hai Traffickers)" 7000
+		add fleet "Smuggler (Hai Traffickers)" 5500
 	system "Mirach"
-		add fleet "Smuggler (Hai Traffickers)" 7000
+		add fleet "Smuggler (Hai Traffickers)" 5500
```
Same deal, if you've had to end up on this plot line, then you're meant to have to _work_ for the information you find. It should not be easy.

UnorderedSigh replied:
In retrospect, the numbers aren't changed enough to matter anyway. Will revert.

MasterOfGrey commented on data/hai/hai reveal 5 philosophers.txt:
```diff
@@ -588,9 +595,8 @@ mission "Hai Reveal [B07-long] Learn of Keystones"
 	clearance
 	to offer
 		has "Hai Reveal [B06-long] Tarazed Analysis: active"
-	source
-		government "Smuggler (Hai Trafficker)"
 	on offer
+		fail "Hai Reveal [B06-long] Tarazed Analysis"
```
Fix already PR'd #7454

UnorderedSigh replied:
You fixed this after I tested it, so I'll retest this part of the campaign with your fix instead.

MasterOfGrey commented on data/hai/hai reveal 5 philosophers.txt:
```diff
@@ -807,17 +815,87 @@ ship "Centipede" "Centipede (Teeneep Jump)"
 		"Outfits Expansion"
 		`"Biroo" Atomic Steering`
 		"Hai Diamond Regenerator"
-		"Hai Williwaw Cooling" 2
-		"Hai Chasm Batteries" 2
+		"Hai Corundum Regenerator"
+		"Hai Chasm Batteries"
+		"Liquid Nitrogen Cooler"
 		"Boulder Reactor"
 		Brig
-		"Fuel Pod" 3
+		"Fuel Pod" 2
+		Supercapacitor
```
Did it perform badly?

Hurleveur replied:
Looking at it I think it just had a rushed layout (had 10 outfit space remaining) and it would overheat **very** quickly
I think we could justify having some of these modifications (like extra fuel) done to the hull and making those more efficient, in another PR, but for now thats a very good change if you ask me (except if that 900 fuel was really needed but I dont think so)

MasterOfGrey commented on data/hai/hai reveal 5 philosophers.txt:
```diff
@@ -807,17 +815,87 @@ ship "Centipede" "Centipede (Teeneep Jump)"
+		"Heavy Anti-Missile Turret" 2
+		"Jump Drive"
+		"LP072a Battery Pack"
+		"Laser Rifle" 25
+		"Liquid Helium Cooler"
+		"Cooling Ducts"
+		"Orca Plasma Steering"
+		"Orca Plasma Thruster"
+		"Outfits Expansion" 4
+		"Fuel Pod" 2
+		"Plasma Turret" 3
+		"Quantum Keystone"
+		"Stack Core"
+		Torpedo 120
+		"Torpedo Launcher" 4
+	turret "Plasma Turret"
+	turret "Heavy Anti-Missile Turret"
+	turret "Heavy Anti-Missile Turret"
+	turret "Plasma Turret"
+	turret "Plasma Turret"
+
+
+ship "Auxiliary" "Auxiliary (Quantum Keystone)"
+	outfits
+		"Sidewinder Missile Launcher" 2
+		"Sidewinder Missile" 90
+		"Quad Blaster Turret" 2
+		"Heavy Anti-Missile Turret" 2
+		"Catalytic Ramscoop"
+		"Armageddon Core"
+		"LP144a Battery Pack"
+		"S-970 Regenerator" 3
+		"Liquid Helium Cooler"
+		"Large Radar Jammer" 2
+		"X4700 Ion Thruster"
+		"X1700 Ion Thruster"
+		"X5200 Ion Steering"
+		"Jump Drive"
+		"Quantum Keystone"
+
```
See #7456

UnorderedSigh replied:
The variants died quickly, especially the Auxiliary.

UnorderedSigh replied:
I'm moving my updated outfits to your variants in variants.txt.

MasterOfGrey commented on data/hai/hai reveal 5 philosophers.txt:
```diff
@@ -876,15 +956,15 @@ mission "Hai Reveal [B12] Devil-Hide Battle"
 	npc save accompany
 		government "Navy (Oathkeeper)"
 		personality escort heroic opportunistic
-		ship "Cruiser (Plasma Anti-Missile)" "N.S. Peacemaker"
+		ship "Cruiser (Quantum Keystone)" "N.S. Peacemaker"
 	npc save accompany
 		government "Republic"
 		personality timid escort opportunistic
-		ship "Auxiliary (Transport)" "N.S. Braveheart"
+		ship "Auxiliary (Quantum Keystone)" "N.S. Braveheart"
 	npc accompany
 		government "Free Worlds"
 		personality escort heroic opportunistic
-		ship "Dreadnought (Jump)" "F.S. Ironwood"
+		ship "Dreadnought (Quantum Keystone)" "F.S. Ironwood"
```
As above

UnorderedSigh replied:
This was because I moved the variants into "hai reveal 5 philosophers.txt," and the purpose of the incorrect ones variants.txt confused me. I will update the variants.txt versions instead.

MasterOfGrey commented on data/hai/hai reveal 5 philosophers.txt:
```diff
@@ -1078,7 +1158,7 @@ mission "Hai Reveal [B13] Freedom"
 		government "Navy (Oathkeeper)"
 		personality escort heroic opportunistic
-		ship "Cruiser (Plasma Anti-Missile)" "N.S. Peacemaker"
+		ship "Cruiser (Quantum Keystone)" "N.S. Peacemaker"
```
As above

MasterOfGrey commented on data/human/variants.txt:
```diff
@@ -1053,7 +1053,6 @@ ship "Cruiser" "Cruiser (Mark II)"
 	turret "Heavy Anti-Missile Turret"
 
 
-# Used in the Hai Reveal campaign, Danforth pulls strings with the Free Worlds to get some cruisers equipped for deep-range deployment
```
Still is as a non-standard plot specific variant.

MasterOfGrey commented on data/human/variants.txt:
```diff
@@ -1073,7 +1072,6 @@ ship "Cruiser" "Cruiser (Plasma Anti-Missile)"
 		"LP144a Battery Pack"
 
 
-
```
3 lines is standard between variants of different ships, 2 between variants of the same ship

UnorderedSigh replied:
I did that by accident while editing. Will revert.

MasterOfGrey reviewed (changes requested):
Many, many good catches and hugely appreciated!

However, a solid range of things that shouldn't be done the way they've been done. I have commented on these individually.

UnorderedSigh commented:
> Holy mother of jesus. Alrighty, will have a look through this. We knew there'd be things after doing a huge rework to it on a short timeline but wow.

I play-tested this many times for @NomadicVolcano when the campaign was originally created, so I knew what was supposed to happen and what was likely to break.

> A couple of these are already noted with fixes PR'd, so you might find that some conflicts arise in future - in which case I'd favour the fixes already PR'd. This should cover off a lot of stuff though.

I will merge those PRs into this one.

UnorderedSigh commented:
I have changed the PR to resolve reviewer comments.

After replaying the entire Hai Reveal, the same way as before, I found more bugs:

1. Hai Rescue: Yeertle Hint - does not vanish when Hai Leaks Response 6 cease to be active
2. Hai Rescue: Oonem Hint - does not vanish when Hai Leaks Response 6 cease to be active
3. Hai Rescue: Prison RP Hint - does not vanish when Hai Leaks Response 6 cease to be active
4. Hai Rescue: Meroot and Lucy Hint - does not vanish when Hai Leaks Response 6 cease to be active
5. Hai Reveal [B01-A] The Blockade Goes Up - does not always see the completion of A14, so I had to change its "to offer." The condition is functionally equivalent.
6. Hai Reveal [B02] Join the Fleet - "npc accompany" implies "npc save" so the mission is impossible if any of the escorts are killed. It is expected that some will die in the battle.
7. Hai Reveal [B02] Join the Fleet - "escort waiting" does nothing; it is the same as "escort." There is no way to make the NPC fleets wait to rendezvous with you at a specified system.
8. Hai Reveal [B02] Join the Fleet - fleets start far away, so they run out of fuel before they arrive. I've changed the mission so the fleets all start at Sunracer, in orbit of the player's planet.

UnorderedSigh commented:
Github won't let me request a re-review from both @MasterOfGrey and @Hurleveur. I've requested just @MasterOfGrey hoping he'll ask a review from whoever else he wants.

MasterOfGrey reviewed (changes requested):
> 1.->4.

In general that's a good catch, they need fail conditions if their missions offer conditions become impossible. Wrong fail conditions though - some of those missions are intentionally still able to be followed up on during the event if you desire. I have suggested the appropriate conditions.

However, this has also alerted me to the fact that the Oonem one should be locked out once the information blackout happens. If she's still on Stormhold when you end up invading the planet, she'd simply rejoin the Hai there and you'd be none the wiser.
Up to you if you want to do that - I'll do it in a bit otherwise.

> 5. Hai Reveal [B01-A] The Blockade Goes Up - does not always see the completion of A14, so I had to change its "to offer." The condition is functionally equivalent.

It doesn't look for the completion of A14, it looks for the offering - but I get what you mean, this is valid.

> 6. Hai Reveal [B02] Join the Fleet - "npc accompany" implies "npc save" so the mission is impossible if any of the escorts are killed. It is expected that some will die in the battle.

This is no longer true. `accompany` and `save` have been divorced from each other (very) recently. These need to have `accompany`.

> 7. Hai Reveal [B02] Join the Fleet - "escort waiting" does nothing; it is the same as "escort." There is no way to make the NPC fleets wait to rendezvous with you at a specified system.

Escort makes it show up on your escort list, waiting means it's already in space when you take off. They're functionally different things and there was never an expectation that they would wait for you anywhere. The NPC ships you have changed are all coming from Mountaintop, you are outside the blockade, they are inside it. They cannot start in Mirfak.

You are told to rush to Sumar to rendezvous precisely because there's no way to make them wait there for you.

> 8. Hai Reveal [B02] Join the Fleet - fleets start far away, so they run out of fuel before they arrive. I've changed the mission so the fleets all start at Sunracer, in orbit of the player's planet.

If fleets run out of fuel *when the player rushes to Sumar like they're told to*, then the variants used need fixing.
If the player doesn't do what they're supposed to and they have issues, then the player will have to learn the difficulty of coordinating a bunch of ships when they don't do their job properly, potentially relying on the emergency ramscoop.

If absolutely necessary we can move the rendezvous to Cardax, we just need to remove the manic pirate spam that occurs in that system in order to make it viable.

> You need to keep a reminder to arrive at the spaceport if the player decides to delay the mission.

Also as a side-note, I might have mixed something up when looking at your changes to the Information Blackout. Your original changes were probably correct, I think I misconstrued them in the context of other proposed changes. You can put those back in if you'd like. Apologies about that.

UnorderedSigh commented on data/hai/hai reveal 5 philosophers.txt:
```diff
@@ -123,20 +123,20 @@ mission "Hai Reveal [B02] Join the Fleet"
+		personality escort heroic opportunistic
 		fleet "Human Wormhole Guards" 20
-	npc accompany
+	npc
 		government "Hai"
-		system "Ultima Thule"
-		personality escort heroic opportunistic waiting
+		system "Mirfak"
+		personality escort heroic opportunistic
 		fleet "Hai Wormhole Guards" 20
-	npc accompany
+	npc
 		government "Hai (Friendly Unfettered)"
-		system "Ultima Thule"
-		personality escort heroic opportunistic vindictive waiting
+		system "Mirfak"
+		personality escort heroic opportunistic vindictive
```
In my tests, if it has "accompany," and any ships in the fleet die, the mission is never completed.

Hurleveur replied:
I think its because we're not on the last version?
Accompany shouldnt have save in it soon if it isnt already the case, I think

MasterOfGrey replied:
> In my tests, if it has "accompany," and any ships in the fleet die, the mission is never completed.

That would have been the case on your initial test because the feature was added and removed for 9.16, and then re-added after due to some issues with the first version.

UnorderedSigh replied:
I got the fleet escort logic to work. The solution is to have the ships spawn at Rajak. Story-wise, they started at Ultima Thule and jumped one system while the player entered orbit.  Now the player is told to rendezvous at Moktar, which they can do even with a slow ship if they hurry. That system isn't too busy. The "various fleets are arriving" dialog at Moktar may not be right, depending on how slow they are, but the fleets should arrive very soon after if they haven't already.

I can't put "accompany" back in because the mission still fails sometimes when ships die. Maybe you could submit an issue for that?

UnorderedSigh commented:
I understand your goal now in the Stormhold invasion. It's a puzzle of sorts: convince the AI escorts to go to the right system all at the same time.

In my tests, "npc accompany" fails if a ship in the fleet does not survive to the end of the mission.

If accompany and save are separated correctly, then I suspect it happens when an entirely fleet is destroyed (like all Syndicate ships). The completion condition doesn't care if the fleet arrived in the *system*; it only cares if they *landed at the destination*. A fleet with no ships doesn't exist in any system, so it cannot have accompanied you to the destination. 

> If fleets run out of fuel when the player rushes to Sumar like they're told to, then the variants used need fixing.

I'm pretty sure they don't all have 600 fuel. They can all make it to Mirfak though (500 fuel) where they can refuel for the remainder of the trip.

> You are told to rush to Sumar

This is not a good place for it. If you go to Sumar, and some of the escorts are past there already, they'll waste fuel jumping back to meet you. Even 600 fuel won't be enough.

Moktar is a better option. It's nearly impossible for the fleet to pass you if you wait there, since it has to be in the path of a fleet that lacks Jump Drives. Also, it has relatively little pirate activity.

Either won't work though because mandating a waypoint doesn't make sense once the player has already won the battle. Sure, they didn't go where you wanted, but the fleet got there anyway and invaded stormhold, there's no reason to fly back to a waypoint with your victorious fleet.

UnorderedSigh commented:
> In general that's a good catch, they need fail conditions if their missions offer conditions become impossible. Wrong fail conditions though - some of those missions are intentionally still able to be followed up on during the event if you desire. I have suggested the appropriate conditions.

I was unable to complete some of those since the mission that followed the hint could not trigger. My assumption was that the mission sequence was supposed to be unavailable. I'll make the opposite change, and ensure the mission sequences can be completed (except the one you said shouldn't be).

I also noticed you don't get information after the first smuggler boarding. Something is wrong with those mission triggers.

UnorderedSigh commented:
I should have time to work on this around 20-ish hours from now. If you merge it before then, I'll open another PR with my remaining fixes.

MasterOfGrey commented:
> I was unable to complete some of those since the mission that followed the hint could not trigger. 

That's why I suggested the appropriate changes to match triggers.

> I'll make the opposite change, and ensure the mission sequences can be completed (except the one you said shouldn't be).

Please just use the suggestions for this case.

UnorderedSigh commented on data/hai/hai reveal 5 philosophers.txt:
```diff
@@ -814,17 +820,19 @@ ship "Centipede" "Centipede (Teeneep Jump)"
 		"Pulse Rifle" 131
 
 
-
```
I'm not sure where this was, so I'm guessing it was "Hai Reveal [B12] Devil-Hide Battle" when I edited the "Centipede (Teeneep Jump)." There's another line before that mission now.

UnorderedSigh commented:
With this third set of changes, I think the campaign works.

1. Better fixes from @MasterOfGrey for smuggler missions.
2. Hai Reveal [A07] Deliver the Relays - workaround for a bug I can't fix. The relays remain in the ship after they're "deployed" so the mission does not complete if the escort with the (now deployed) relays hasn't arrived. Now the mission description tells you to return the "deployment hardware" and I've left a note in a comment that we need a better fix.
3. Hai Reveal [B01] Information Blackout - reminder to visit the spaceport remains, after Master of Grey approved me putting it back in.
4. Hai Reveal [B02] Join the Fleet - corrected mission name
5. Hai Reveal [B02] Join the Fleet - ships spawn at Rajak so they have enough fuel to reach Stormhold. There is a comment explaining this.
6. Hai Reveal [B02] Join the Fleet - player is told to meet fleet at Moktar. This ensures the fleet doesn't pass the player and have to jump back, running out of fuel. 
7. Hai Reveal [B02] Join the Fleet - the "wow that's a big fleet" message is tweaked to make sense if the fleet hasn't arrived yet, since the prior sentence says you see their transpoder signals
8. "Thankyou" is not a word
9. Hai Reveal [B06-long A] Smuggler Ambush - mission is able to trigger now
10. Hai Reveal [B06-long] Tarazed Analysis - other half of bug fix to B06-long A 

I found these problems which I did not fix, but they don't stop the campaign:

1. Hai Reveal [A14] A Time of Trouble - escort randomly-chosen ships that may not have enough fuel. This seams reasonable to me; this is *not* an early game escort mission. Helping a ship with too little fuel is an okay added challenge. 
2. Pincer Beasts cannot keep up with asteroids in some systems. They just wiggle around trying to keep up until they get bored and leave. Reverse thrusters, and stronger forward thrusters would help.
3. Hai Reveal [B02] Join the Fleet - fails when a ship with "npc accompany" dies, even if it does not have "save." Something in the engine isn't working how it should. My workaround is to remove "accompany" until someone fixes this.

MasterOfGrey reviewed (approved):
This is in a good place to merge in.

Anything else should be addressed separately after this is in place.

UnorderedSigh commented:
> This is in a good place to merge in.

Excellent! My first successful Endless Sky PR.

> Anything else should be addressed separately after this is in place.

A wise decision. Two of the remaining problems are engine trouble, which may take a while to fix.

Hurleveur commented:
Also is smth that should be addressed in another PR: 3k battle experience required for HR part4 and its not even mentionned to train for battle or smth.

Hurleveur reviewed (approved):
Took a look at the changes, everything is sound to me, it does the issues I noticed when playing trough HR

UnorderedSigh commented:
> Also is smth that should be addressed in another PR: 3k battle experience required for HR part4 and its not even mentionned to train for battle or smth.

To get to part 4, you have to finish the entire human campaign. Can you finish that with low experience?

UnorderedSigh commented:
Now that the campaign actually can be played through, we'll get more testers and bug reports.

There should be more autosaves and "on visit" messages too. Without that, I foresee people getting confused by mission objectives, but since I've played the campaign at least 20 times, I don't know where that'll happen.

Hurleveur commented:
> > Also is smth that should be addressed in another PR: 3k battle experience required for HR part4 and its not even mentionned to train for battle or smth.
> 
> To get to part 4, you have to finish the entire human campaign. Can you finish that with low experience?

Indeed. I played trough the entire campaign and got a lot of mercenary missions, really a lot

UnorderedSigh commented:
In that case, I don't think you need 3k to finish it. The Stormhold battle isn't that hard. Even Devil Hide should not be any harder than the most difficult main campaign battles.

That definitely belongs in a separate PR though, because we need a wider discussion on what the combat requirement should be, if there are any at all.

Hurleveur commented:
Indeed, as thats linked to the tribute, I opened a PR that addresses that a bit among other things.

In the meantime this should be merged ASAP as HR is not currently playable without those fixes.

MasterOfGrey commented:
If @W1zrad, @Amazinite or @Anarchist2 could push this one through sooner rather than later that would be greatly appreciated so we can push on with some things that are waiting on it.

UnorderedSigh commented:
You cannot reach the Wanderers storyline without these fixes. The bugs lock out about a third of the major stories in the game.
